pkg/remote/write.go

```diff
@@ -0,0 +1,252 @@
+package remote
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+
+	mcontext "github.com/grafana/prometheus-pulsar-remote-write/pkg/context"
+	"github.com/grafana/prometheus-pulsar-remote-write/pkg/pulsar"
+	"github.com/grafana/prometheus-pulsar-remote-write/pkg/version"
+)
+
+type Write struct {
+	logger log.Logger
+
+	samplePerTenantID   map[string][]pulsar.ReceivedSample
+	deadlinePerTenantID map[string]time.Time
+
+	// BatchSize is the amount of samples that are aggregated into a single
+	// remote write request
+	BatchSize int
+	// BatchMaxDelay is the maximum delay acceptable for a batch to wait to
+	// reach the BatchSize
+	BatchMaxDelay time.Duration
+
+	// checkInterval limits how often we want to check on send conditions for batches are met
+	checkInterval time.Duration
+}
+
+type WriteOpts func(o *Write)
+
+func WithLogger(l log.Logger) WriteOpts {
+	return func(w *Write) {
+		w.logger = l
+	}
+}
+
+func NewWrite(opts ...WriteOpts) *Write {
+	w := &Write{
+		logger: log.NewNopLogger(),
+
+		samplePerTenantID:   make(map[string][]pulsar.ReceivedSample),
+		deadlinePerTenantID: make(map[string]time.Time),
+
+		BatchSize:     100,
+		BatchMaxDelay: 5 * time.Second,
+		checkInterval: 100 * time.Millisecond,
+	}
+
+	for _, opt := range opts {
+		opt(w)
+	}
+
+	return w
+}
+
+type headerRoundtripper struct {
+	upstream http.RoundTripper
+}
+
+func (t *headerRoundtripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// if tenant ID is set, expose it in the header
+	tenantID := mcontext.TenantIDFromContext(req.Context())
+	if tenantID != "" {
+		req.Header.Set(mcontext.HTTPHeaderTenantID, tenantID)
+	}
+
+	// overwrite the user agent header
+	req.Header.Set("User-Agent", fmt.Sprintf("%s/%s", version.AppName(), version.AppVersion()))
+
+	return t.upstream.RoundTrip(req)
+}
+
+type ClientConfig = remote.ClientConfig
+
+func NewWriteClient(conf *ClientConfig) (remote.WriteClient, error) {
+	clientInterface, err := remote.NewWriteClient("remote-write", conf)
+	if err != nil {
+		return nil, err
+	}
+
+	client, ok := clientInterface.(*remote.Client)
+	if !ok {
+		return nil, fmt.Errorf("client of unexpected type %T", clientInterface)
+	}
+
+	// add custom rountripper to modify user-agent and tenant header if necessary
+	client.Client.Transport = &headerRoundtripper{upstream: client.Client.Transport}
+
+	return client, nil
+}
+
+func (w *Write) Run(ctx context.Context, sampleCh chan pulsar.ReceivedSample, client remote.WriteClient) error {
+	tick := time.NewTicker(w.checkInterval)
+	defer tick.Stop()
+
+	// this is set when a retry able error happend
+	errRemoteWriteRetryable := false
+	blockingSampleCh := make(chan pulsar.ReceivedSample)
```

Comment:
I'm not sure I understand the purpose of the `blockingSampleCh`...

So when a retryable error happens on remote write then:
1) `errRemoteWriteRetryable` gets set to `true` on `:182`
2) on the next iteration `sample()` gets called again on `:131`
3) because `errRemoteWriteRetryable` is `true`, `sample()` returns `blockingSampleCh`
4) the `select` at the beginning of the loop will then wait for `blockingSampleCh` to yield an object or the ticker to tick or the `ctx` to get cancelled
5) most likely the next event will then be the ticker ticking, causing a retry

So is the desired function of `blockingSampleCh` to introduce a short wait time between retries? If that's the case, then wouldn't we only want to sleep before sending the next sample of that tenant for which the error occurred (since limits or often tenant specific)? In the current implementation, wouldn't that sleep block all tenants?

Reply:
Actually, even if it's true that we're blocking the forwarding of data for all tenants, instead of just blocking it for the ones which yielded errors, i don't think this is an issue which is critical enough to block the merging of this PR. so i'll just approve anyway to not unnecessarily block you

Reply:
Very good point you are raising here. Currently when an error for only a single tenant happens on the remote write end of the adapter we would still receive messages from pulsar for that tenant as they are all coming from the same queue and we have no way to tell the tenant.

I think there would be more intelligent was of handling that e.g.:
- `Nack` messages for blocked tenants, but that would mean we need to have a good resumption strategy to avoid out of order samples. 
- Having different queue per tenant in pulsar

For now I would like to keep that simple behaviour and figure out what would be the best way forward in terms of real world error cases. 

I have improved the comment in code to reflect that